Code-completion, indexing, cursor-info, etc. for KeyPath dynamic member lookup

benlangmuir opened the pull request:
Add support for code-completion looking up keypath dynamic members, and cursor-info/indexing walking the underlying decls.

rdar://49028783
rdar://49028895
rdar://49029126

benlangmuir committed:
[index] Add indexing of references in keypath member lookup expressions

Currently includes both the implicit references to the
subscript(dynamicMember:) and the explicit references to the underlying
property/subscript declarations.

rdar://49028783

benlangmuir committed:
[index] Fix dynamicMemberLookup subscript reference implicit role

When building the implicit subscript expression, set the "implicit" bit
correctly and pass it through in the indexer so that we get implicit
refernces to the subscript. This would be useful for e.g. searching for
all uses of the dynamic subscript.

benlangmuir committed:
Handle dynamic member lookup in annotation and cursor info

Ensure the various entity walkers handle the implicit subscript
reference correctly (usually by ignoring it) and fall through to the
underlying declarations.

rdar://49028895

benlangmuir committed:
[code-completion] Add completion for keypath dynamic member lookup

Looks into the root type of the keypath to find additional members. This
does not currently map the type of the completion to the subscript's
return type.

rdar://49029126

benlangmuir commented:
@swift-ci please smoke test

rintaro commented on lib/Sema/LookupVisibleDecls.cpp:
```diff
@@ -621,6 +621,41 @@ static void lookupVisibleMemberDeclsImpl(
+
+    auto subs =
+        baseType->getMemberSubstitutionMap(dc->getParentModule(), subscript);
+    if (auto memberType = rootType->subst(subs))
```
`if (memberType->mayHaveMembers())` just in case.

rintaro commented on lib/Sema/LookupVisibleDecls.cpp:
```diff
@@ -839,6 +886,14 @@ static void lookupVisibleMemberDecls(
 
+  llvm::DenseSet<DeclBaseName> knownMembers;
+  for (auto &kv : overrideConsumer.FoundDecls) {
+    knownMembers.insert(kv.first);
```
Could we avoid this copy if unless it's `hasDynamicMemberLookupAttribute()`?

benlangmuir replied:
Good idea.

benlangmuir replied:
Fixed in my latest changes.  Also added the check for `mayHaveMembers`.

rintaro commented on lib/Sema/LookupVisibleDecls.cpp:
```diff
@@ -822,6 +857,18 @@ class OverrideFilteringConsumer : public VisibleDeclConsumer {
+    // Dynamic lookup members are only visible if they are not shadowed by
+    // non-dynamic members.
+    if (seen.count(VD->getBaseName()) == 0)
+      consumer.foundDecl(VD, reason);
```
The `reason` here is necessarily accurate. Maybe, we should introduce `DeclVisibilityKind::DynamicMember`?
Also we might want to propagate `subscript` and `memberType` to the consumer, somehow 🤔

rintaro replied:
Let's pursue this in follow-up PRs.

benlangmuir replied:
I thought about a new `DeclVisibilityKind`, but hesitated because it's orthogonal from some of the existing cases:
* `MemberOfCurrentNominal`
* `MemberOfProtocolImplementedByCurrentNominal`
* `MemberOfSuper`

Ideally we could distinguish dynamic member of current nominal vs. dynamic member of super.  It's also not clear to me how to prioritize dynamic members vs regular members - does it affect priority? If so, how?

We could add three new kinds, or possibly add a separate bit, I'm not sure.

---

I agree with the overall point that we probably want to add this new information in order to handle the type mapping correctly in the future.  Do you think it would be sufficient for the lookup code to determine a type using the subscript + memberType and pass that single type back to code-completion, or do we want to bundle up all the pieces so  that code-completion can also add the contextual type before solving?

benlangmuir committed:
Update index_keypath_member_lookup for change to implicit subscript location

The implicit subscript is not located at the "." instead of the
identifier, so update the index test.

benlangmuir committed:
Remove unnecessary check from cursor info keypath member lookup test

This test baked in an incorrect USR from a bug that has since been
fixed. But we don't really need to check this anyway, so remove it.

benlangmuir committed:
[code-completion] Avoid name copy for non dynamic member types

Per review feedback. Also add some test cases that should fail dynamic
member lookup.

benlangmuir commented:
@swift-ci please test

benlangmuir commented:
@swift-ci please test Linux

nathawes commented on test/Index/index_keypath_member_lookup.swift:
```diff
@@ -0,0 +1,139 @@
+// CHECK: [[WBR_LINE:[0-9]+]]:17 | instance-property/Swift | y | [[PY_USR]] | Ref,Writ,RelCont | rel: 1
+// CHECK: [[WBR_LINE:[0-9]+]]:17 | instance-method/acc-set/Swift | setter:y | [[PY_SET_USR]] | Ref,Call,Impl,RelCall,RelCont | rel: 1
+
+  // FIXME: crashes typechecker rdar://problem/49533404
```
From that radar, it looks like Pavel fixed this already.

benlangmuir replied:
Ah, forgot I still had this FIXME.  I'll take care of this in a follow-up.

nathawes reviewed (approved):
The indexing part LGTM (Rintaro asked if I could take a look).

benlangmuir commented:
Foundation Linux failure is unrelated (again).
